tf_helpers/tf_dot_general.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""
+Construct an equivalent general dot operation as that in JAX -
+    <https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.dot_general.html>
+
+Although there is an implementation in TF XLA, avoid directly using XLA when
+possible.
+
+Zhibo Zhang, 2020.06.30
+"""
+
+import tensorflow as tf
+from tensorflow.python.ops import numpy_ops as tf_np
+import string
+
+
+def _minus(a, b):
+  return [x for x in a if x not in b]
+
+
+def compose_output_rep(lhs_rep, rhs_rep, lhs_contraction, rhs_contraction,
+                        lhs_batch, rhs_batch):
+  """ Compose the output string representation.
+
+  e.g., ij, jk, (((1,), (0,)), ((), ())) -> ik
+        aij, ajk, (((2,), (1,)), ((0,), (0,))) -> aik
+
+  Args:
+    lhs_rep: A string representation for the left-hand side input array
```

Comment:
Here and elsewhere, I'd suggest using type annotations (e.g. as in https://jax.readthedocs.io/en/latest/_modules/jax/lax/lax.html) instead of specifying types in the docstring. Feel free to postpone this until later if we're short on time.